resources/net.go

```diff
@@ -0,0 +1,804 @@
+
+		case event, ok := <-obj.Events():
+			if !ok {
+				closeMap[4] = true // done
+				continue
+			}
+			if exit, send = obj.ReadEvent(event); exit != nil {
+				return *exit // exit
+			}
+		}
+
+		// do all our event sending all together to avoid duplicate msgs
+		if send {
+			send = false
+			obj.Event()
+		}
+	}
+}
+
+// ifaceCheckApply checks the state of the network device and performs the
+// appropriate actions to converge its state with the graph.
+func (obj *NetRes) ifaceCheckApply(apply bool) (checkOK bool, _ error) {
+	var ok bool
+
+	// create a bus object pointing to the adapter
+	busObj := obj.gdbus.Object(dBusNetworkdInterface, obj.ifacePath)
+	// get the object's operational state
+	call := busObj.Call(dBusPropertiesGet, 0, dBusNetworkdLinkInterface, opState)
+	if call.Err != nil {
+		return false, call.Err
+	}
+	// check the operational state
+	os, ok := call.Body[0].(dbus.Variant)
+	if !ok {
+		return false, fmt.Errorf("error casting opstate")
+	}
+
+	// if the interface's state is correct, we're done
+	if obj.State == ifaceDown && os == dbus.MakeVariant(opStateOff) {
+		return true, nil
+	}
+	if obj.State == ifaceUp {
+		if os != dbus.MakeVariant(opStateOff) {
+			return true, nil
+		}
+		// XXX: retry logic for fault tolerance
+		if os == dbus.MakeVariant(opStateNoCarrier) {
+			return false, fmt.Errorf("connection failed")
+		}
+	}
+
+	// if the service is restarting, wait for the next event
+	restart, err := obj.svcRestarting()
+	if err != nil {
+		return false, err
+	}
+	if restart {
+		return false, nil
+	}
+
+	if !apply {
+		return false, nil
+	}
+
+	log.Printf("%s: ifaceCheckApply(%t)", obj, apply)
+
+	// enable or disable the interface
+	cmd := exec.Command("ip", "link", "set", obj.GetName(), obj.State)
+
+	// open a pipe to get error messages from os/exec
+	stderr, err := cmd.StderrPipe()
+	if err != nil {
+		return false, errwrap.Wrapf(err, "failed to initialize stderr pipe")
+	}
+
+	// start the command
+	if err := cmd.Start(); err != nil {
+		return false, errwrap.Wrapf(err, "cmd failed to start")
+	}
+	// capture any error messages
+	slurp, err := ioutil.ReadAll(stderr)
+	if err != nil {
+		return false, errwrap.Wrapf(err, "error slurping error message")
+	}
+	// wait until cmd exits and return error message if any
+	if err := cmd.Wait(); err != nil {
+		return false, errwrap.Wrapf(err, "%s", slurp)
+	}
+	// XXX: wait for state?
+
+	return false, nil
+}
+
+// fileCheckApply checks and maintains the systemd-networkd unit file contents.
+// TODO: check for conflicting unit files and think about unit priority
+func (obj *NetRes) fileCheckApply(apply bool) (checkOK bool, _ error) {
+	// check if the unit file exists
+	_, err := os.Stat(obj.unitFilePath)
+	if err != nil {
+		if !os.IsNotExist(err) {
+			return false, errwrap.Wrapf(err, "error checking file")
+		}
+	}
+	// build the unit file contents from the definition
+	contents := obj.unitFileContents()
+	// check the file contents
+	if err == nil {
```

Comment:
I think you can do a err != nil above, and then inline this? please look into it